### ISSUE TITLE
TestGracefulBrowserShutdown failed in headless-shell since 99.0.4818.2

### ISSUE THREAD

**Issue by ZekeLu** [chromedp/chromedp#1078]
See https://github.com/chromedp/chromedp/actions/runs/2430302280. This is the first run after `chromedp/headless-shell:latest` is updated to `102.0.5005.63` (it was `98.0.4758.102` before). A test shows it failes since the very first release of v99 (namely, `99.0.4818.2`).

`TestGracefulBrowserShutdown` tests the graceful shutdown by checking the cookie surviving the restart of the browser. It seems that the cookie is not persisted since `chromedp/headless-shell:99.0.4818.2`, which causes the test to fail.

Two things to do next:
1. do a bisect to find out the commit that introduced this change;
2. skip the test in headless-shell or modify the implementation.

**Comment by ZekeLu** [comment]
### The Cause

The bisect shows that the first bad commit is: https://github.com/chromium/chromium/commit/f95c449bb79a961ae3332f6783f770159a3e1189

<details>
<summary>changes in the commit that introduced the bug</summary>

```diff
diff --git a/headless/lib/browser/headless_request_context_manager.cc b/headless/lib/browser/headless_request_context_manager.cc
index 7a3307be2b513..46f0aaeefe5cb 100644
--- a/headless/lib/browser/headless_request_context_manager.cc
+++ b/headless/lib/browser/headless_request_context_manager.cc
@@ -68,7 +68,8 @@ void SetCryptKeyOnce(const base::FilePath& user_data_path) {
 #if (defined(OS_WIN) || defined(OS_LINUX)) && defined(HEADLESS_USE_PREFS)
   // The OSCrypt keys are process bound, so if network service is out of
   // process, send it the required key if it is available.
-  if (content::IsOutOfProcessNetworkService()) {
+  if (content::IsOutOfProcessNetworkService() &&
+      OSCrypt::IsEncryptionAvailable()) {
     content::GetNetworkService()->SetEncryptionKey(
         OSCrypt::GetRawEncryptionKey());
   }
```
</details>

After this change, **the cookies are not saved to the disk**. According to https://storage.googleapis.com/chromium-find-releases-static/f95.html#f95c449bb79a961ae3332f6783f770159a3e1189, the commit was initially landed in `99.0.4781.0`.

### The Fix

And it's fixed later by: https://github.com/chromium/chromium/commit/590e02c7b515fa80862053b503cb7473e8786566

<details>
<summary>changes that fixed the bug</summary>

```diff
diff --git a/headless/lib/browser/headless_request_context_manager.cc b/headless/lib/browser/headless_request_context_manager.cc
index 46f0aaeefe5cb..36358942d2f07 100644
--- a/headless/lib/browser/headless_request_context_manager.cc
+++ b/headless/lib/browser/headless_request_context_manager.cc
@@ -68,8 +68,11 @@ void SetCryptKeyOnce(const base::FilePath& user_data_path) {
 #if (defined(OS_WIN) || defined(OS_LINUX)) && defined(HEADLESS_USE_PREFS)
   // The OSCrypt keys are process bound, so if network service is out of
   // process, send it the required key if it is available.
-  if (content::IsOutOfProcessNetworkService() &&
-      OSCrypt::IsEncryptionAvailable()) {
+  if (content::IsOutOfProcessNetworkService()
+#if defined(OS_WIN)
+      && OSCrypt::IsEncryptionAvailable()
+#endif
+  ) {
     content::GetNetworkService()->SetEncryptionKey(
         OSCrypt::GetRawEncryptionKey());
   }
```
</details>

According to https://storage.googleapis.com/chromium-find-releases-static/590.html#590e02c7b515fa80862053b503cb7473e8786566, the fix was initially landed in `99.0.4808.0`.

And I have tested the following versions of `headless_shell` (compiled from the source code directly), all of them work as expected (the cookie is correctly saved and can be read back later):
- [x] [99.0.4818.2](https://github.com/chromium/chromium/tree/99.0.4818.2)
- [x] [100.0.4896.56](https://github.com/chromium/chromium/tree/100.0.4896.56)
- [x] [101.0.4947.3](https://github.com/chromium/chromium/tree/101.0.4947.3)
- [x] [102.0.5005.63](https://github.com/chromium/chromium/tree/102.0.5005.63)
- [x] [103.0.5060.33](https://github.com/chromium/chromium/tree/103.0.5060.33)
- [x] [104.0.5083.0](https://github.com/chromium/chromium/tree/104.0.5083.0)

<details>
<summary><code>out/Default/args.gn</code> used to compile <code>headless_shell</code></summary>

```gn
symbol_level=0
blink_symbol_level=0
v8_symbol_level=0
is_debug=false
dcheck_always_on=false
is_official_build=true
chrome_pgo_phase=0
```
</details>

<details>
<summary>app used to test <code>headless_shell</code></summary>

```go
package main

import (
	"context"
	"flag"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/http/httptest"
	"os"
	"time"

	"github.com/chromedp/chromedp"
)

func main() {
	addr := flag.String("l", ":8080", "the address to listen for HTTP request")
	userDataDir := flag.String("d", "/tmp/chromedp-test", "the user data directory passed to --user-data-dir")
	mode := flag.Int("m", 3, "testing mode. 1: set cookie, 2: get cookie, 3: do both 1&2")
	execPath := flag.String("p", "./headless_shell", "The path to the headless_shell")
	flag.Parse()

	// start a test server
	ln, err := net.Listen("tcp", *addr)
	if err != nil {
		log.Fatal(err)
	}

	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.RequestURI == "/set" {
			http.SetCookie(w, &http.Cookie{
				Name:    "cookie1",
				Value:   "value1",
				Expires: time.Now().AddDate(0, 0, 1), // one day later
			})
		}
	}))
	defer ts.Close()
	ts.Listener = ln
	ts.Start()

	log.Println("web server:", ts.URL)

	if *userDataDir == "" {
		dir, err := ioutil.TempDir("", "chromedp-test")
		if err != nil {
			log.Fatal(err)
		}
		defer os.RemoveAll(dir)

		*userDataDir = dir
		log.Println("temp directory created:", *userDataDir)
	}
	log.Println("user data directory:", *userDataDir)

	opts := []chromedp.ExecAllocatorOption{
		chromedp.NoFirstRun,
		chromedp.NoDefaultBrowserCheck,
		chromedp.Headless,
		chromedp.UserDataDir(*userDataDir),
		chromedp.ExecPath(*execPath),
	}
	actx, cancel := chromedp.NewExecAllocator(context.Background(), opts...)
	defer cancel()

	// set cookies
	if (*mode)&0x01 > 0 {
		log.Println("about to start a new browser to set cookies")
		ctx, _ := chromedp.NewContext(actx)
		if err := chromedp.Run(ctx, chromedp.Navigate(ts.URL+"/set")); err != nil {
			panic(err)
		}

		if err := chromedp.Cancel(ctx); err != nil {
			panic(err)
		}
		log.Printf("cookie set. Run 'sqlite3 %s/Default/Cookies .dump' to check the saved cookies\n", *userDataDir)
		log.Println("the browser is closed")
	}
	// get cookies
	if (*mode)&0x02 > 0 {
		log.Println("about to start a new browser to read cookies")
		ctx, _ := chromedp.NewContext(actx)
		var got string
		if err := chromedp.Run(ctx,
			chromedp.Navigate(ts.URL),
			chromedp.EvaluateAsDevTools("document.cookie", &got),
		); err != nil {
			panic(err)
		}
		if want := "cookie1=value1"; got != want {
			log.Printf("failed. Want cookies %q; got %q\n", want, got)
			os.Exit(1)
		}
		log.Printf("succeeded in getting cookies: %q\n", got)
	}
}
```
</details>

### There Is Something Wrong with the `chromedp/headless-shell` Image

But the test failed in all the corresponding `chromedp/headless-shell` images:
- [ ] chromedp/headless-shell:99.0.4818.2
- [ ] chromedp/headless-shell:100.0.4896.56
- [ ] chromedp/headless-shell:101.0.4947.3
- [ ] chromedp/headless-shell:102.0.5005.63
- [ ] chromedp/headless-shell:103.0.5060.33
- [ ] chromedp/headless-shell:104.0.5083.0

<details>
<summary>Run <code>./contrib/docker-test.sh</code> to test <code>chromedp/headless-shell</code></summary>

```diff
diff --git a/contrib/docker-test.sh b/contrib/docker-test.sh
index cefc03b..551fa10 100755
--- a/contrib/docker-test.sh
+++ b/contrib/docker-test.sh
@@ -4,7 +4,7 @@ SRC=$(realpath $(cd -P "$(dirname "${BASH_SOURCE[0]}")" && pwd)/..)
 
 pushd $SRC &> /dev/null
 
-IMAGE=${IMAGE:-chromedp/headless-shell:latest}
+IMAGE=${IMAGE:-chromedp/headless-shell:104.0.5083.0}
 
 set -e
 
@@ -20,7 +20,7 @@ set -e
     --workdir=/chromedp \
     --env=PATH=/headless-shell \
     --env=HEADLESS_SHELL=1 \
-    $IMAGE -test.v -test.parallel=1 -test.timeout=3m
+    $IMAGE -test.v -test.parallel=1 -test.timeout=3m -test.run=^TestGracefulBrowserShutdown$
 )
 
 popd &> /dev/null
```
</details>

### Next Step

Maybe `defined(OS_WIN)` (or `BUILDFLAG(IS_WIN)`) return an incorrect value on the builder (I don't know how can it be). We can try to apply the following changes to [104.0.5083.0](https://github.com/chromium/chromium/tree/104.0.5083.0) to figure out whether this is the culprit.

```diff
diff --git a/headless/lib/browser/headless_request_context_manager.cc b/headless/lib/browser/headless_request_context_manager.cc
index 60c331241dfab..3ef248dc64bbb 100644
--- a/headless/lib/browser/headless_request_context_manager.cc
+++ b/headless/lib/browser/headless_request_context_manager.cc
@@ -69,9 +69,6 @@ void SetCryptKeyOnce(const base::FilePath& user_data_path) {
   // The OSCrypt keys are process bound, so if network service is out of
   // process, send it the required key if it is available.
   if (content::IsOutOfProcessNetworkService()
-#if BUILDFLAG(IS_WIN)
-      && OSCrypt::IsEncryptionAvailable()
-#endif
   ) {
     content::GetNetworkService()->SetEncryptionKey(
         OSCrypt::GetRawEncryptionKey());
```

@kenshaw Can you try that on the builder?

**Comment by ZekeLu** [comment]
My bad! I forgot to import `build/args/headless.gn`.

Here is the updated `out/Default/args.gn` file used to compile `headless_shell` for the testing:

```gn
import("//build/args/headless.gn")
symbol_level=0
blink_symbol_level=0
v8_symbol_level=0
is_debug=false
dcheck_always_on=false
is_official_build=true
chrome_pgo_phase=0
```

It turns out that the breaking change is introduced by https://github.com/chromium/chromium/commit/7c64b00ce8ba4f984b8a48cf2a62025a9bf2fa48. According to https://storage.googleapis.com/chromium-find-releases-static/7c6.html#7c64b00ce8ba4f984b8a48cf2a62025a9bf2fa48, this commit was initially landed in `99.0.4772.0`.

This is a breaking change instead of a bug because it seems that this is an expected behavior according to the following commits:

- https://github.com/chromium/chromium/commit/abb345ff362461181eff5ec68e65c370da516029 (initially landed in 89.0.4351.0)
- https://github.com/chromium/chromium/commit/9250cb6111ebfa5f751141ae5c5443552b097cf2 (initially landed in 98.0.4746.0)

https://github.com/chromium/chromium/commit/7c64b00ce8ba4f984b8a48cf2a62025a9bf2fa48 just (unintentionally?) uncovers this behavior.

The fix is easy, just adding `headless_use_prefs = true` to `args.gn` to change this behavior.
